### DISCUSSION TITLE
overlapping plot in time domain simulation when Fault model added

### DISCUSSION THREAD

**Discussion by seedpower** [CURENT/andes#384] [Q&A]
```
import numpy as np
import andes

ss = andes.run(andes.get_case("ieee14/ieee14_full.xlsx"), setup=False, default_config=True)
ss.add("Fault", dict(bus=9, tf=0.4, tc=0.7))
ss.setup()
ss.PFlow.run()

ss.TDS.config.tstep = 0.1
ss.TDS.config.shrinkt = 1

for tf in np.linspace(0.1, 1.2, 12):
    ss.TDS.config.tf = tf
    ss.TDS.run()

# ss.TDS.config.tf = 1.2
# ss.TDS.run()

ss.TDS.load_plotter()
ss.TDS.plotter.plot(ss.GENROU.delta)
```
The diagram (P1) is:
![image](https://user-images.githubusercontent.com/11022830/202730352-43359cbc-5c07-44dc-9e46-549a8f48c457.png)

Now, I change code to this:
```
# for tf in np.linspace(0.1, 1.2, 12):
#     ss.TDS.config.tf = tf
#     ss.TDS.run()

ss.TDS.config.tf = 1.2
ss.TDS.run()
```
The diagram (P2) is:
![image](https://user-images.githubusercontent.com/11022830/202730568-48227900-98e5-47d6-a4a0-9792aba18090.png)

If I set `ss.TDS.config.shrinkt = 0`，use below code:
```
ss.TDS.config.tstep = 0.1
ss.TDS.config.shrinkt = 0

for tf in np.linspace(0.1, 1.2, 12):
    ss.TDS.config.tf = tf
    ss.TDS.run()

# ss.TDS.config.tf = 1.2
# ss.TDS.run()
```
The diagram (P3) is:
![image](https://user-images.githubusercontent.com/11022830/202731930-30e4beb7-88d4-42ae-849a-915c0f1d0fc5.png)

Existing problems:
1. compare P1 with P2, TDS resume,  overlapping segments `0.4~0.5` and `0.7~0.8`.
2. compare P1 with P3, setting `ss.TDS.config.shrinkt=0` , overlapping segment become shorter.

**Comment by cuihantao** [comment]
Thanks for reporting. I will fix it asap. It might have to do with the timing logic again.

**Comment by cuihantao** [comment]
I pushed a fix to my `develop` branch. Please test and let me know. Thanks!

**Reply by seedpower** [reply]
The diagram looks OK, but a bug is found during operation.

```
-> Single process finished in 0.3824 seconds.
100%|#####################################| 100.0/100 [00:00<00:00, 16681.80%/s]
100%|######################################| 100.0/100 [00:00<00:00, 5878.82%/s]
[1, 3, 4] -4 0.4349480709651291
100%|######################################| 100.0/100 [00:00<00:00, 6252.78%/s]
[1, 3, 5] -4 0.5966469074013996
100%|######################################| 100.0/100 [00:00<00:00, 6248.13%/s]
[] 16 0.7530446875293808
100%|######################################| 100.0/100 [00:00<00:00, 7691.03%/s]
[1, 4] 0 0.7518951434124929
100%|######################################| 100.0/100 [00:00<00:00, 8331.12%/s]
[] 16 0.665426697833909
100%|######################################| 100.0/100 [00:00<00:00, 6248.68%/s]
[] 16 0.4978766214683873
100%|######################################| 100.0/100 [00:00<00:00, 7690.04%/s]
[1, 4] 0 0.40892248816036414
100%|######################################| 100.0/100 [00:00<00:00, 6662.28%/s]
[1, 2] 0 0.4881547919887098
100%|######################################| 100.0/100 [00:00<00:00, 4345.66%/s]
[] 16 0.6355832660542297
<Fault Fault_1>: Applying fault on Bus (idx=15) at t=1.0 sec.
100%|######################################| 100.0/100 [00:00<00:00, 1298.16%/s]
[1, 2, 3] -4 0.6793204247849233
100%|######################################| 100.0/100 [00:00<00:00, 6248.50%/s]
[] 16 0.5761988848454851
100%|######################################| 100.0/100 [00:00<00:00, 5877.42%/s]
[1, 3] 0 0.5012154259710483
100%|######################################| 100.0/100 [00:00<00:00, 5885.34%/s]
[1] 4 0.44951409404627096
100%|######################################| 100.0/100 [00:00<00:00, 4541.01%/s]
[] 16 0.3755133298491368
100%|######################################| 100.0/100 [00:00<00:00, 5550.59%/s]
[] 16 0.3577436445877031
100%|######################################| 100.0/100 [00:00<00:00, 4543.32%/s]
[1] 4 0.3804826139394568
100%|######################################| 100.0/100 [00:00<00:00, 7144.59%/s]
[] 16 0.30324752067766875
<Fault Fault_1>: Clearing fault on Bus (idx=15) at t=1.833333333333333 sec.
 96%|####################################6 | 96.49/100 [00:00<00:00, 7417.40%/s]C:\Users\Administrator\.andes\pycode\ESST3A.py:21: RuntimeWarning: invalid value encountered in sqrt
  return (-v + vbus, LAW2_y*VB_y*ue - vout, -UEL + UEL0, ue*(-IN*VE + KC*XadIfd), -FEX_y + select([less_equal(IN, 0),less_equal(IN, 0.433),less_equal(IN, 0.75),less_equal(IN, 1),greater(IN, 1),__trues], [__ones,1 - 0.577*IN,sqrt(0.75 - IN**2),1.732 - 1.732*IN,__zeros,__zeros], default=nan), FEX_y*VE - VB_x, VBMAX*VB_lim_zu + VB_lim_zi*VB_x - VB_y, KG*vout - VG_x, VGMAX*VG_lim_zu + VG_lim_zi*VG_x - VG_y, LAW1_y - VG_y - vrs, -vref + vref0, -LG_y - vi + vref, HLI_zi*vi + HLI_zl*VIMIN + HLI_zu*VIMAX - vil, HG_lt_z0*UEL + HG_lt_z1*vil - HG_y, LL_LT1_z1*LL_LT2_z1*(-LL_x + LL_y) + LL_x*TB - LL_y*TB + TC*(HG_y - LL_x), ue*(-vf0 + vout), 0, 0, 0, 0, 0, 0, 0)
C:\Users\Administrator\.andes\pycode\ESST3A.py:33: RuntimeWarning: invalid value encountered in sqrt
  return (-1, 1, -1, LAW2_y*ue, -1, -VE*ue, KC*ue, select([less_equal(IN, 0),less_equal(IN, 0.433),less_equal(IN, 0.75),less_equal(IN, 1),__trues], [__zeros,-0.577*__ones,-IN/sqrt(0.75 - IN**2),-1.732*__ones,__zeros], default=nan), -1, VE, -1, VB_lim_zi, -1, KG, -1, VG_lim_zi, -1, -1, -1, -1, 1, -1, HLI_zi, -1, HG_lt_z0, HG_lt_z1, -1, TC, LL_LT1_z1*LL_LT2_z1 - TB, ue)
Traceback (most recent call last):

...

  File "C:\Users\Administrator\Desktop\andes\pycharm-test\env_ieee14.py", line 96, in step
    res = self.ss.TDS.run()
  File "C:\ProgramData\Anaconda3\lib\site-packages\andes\routines\tds.py", line 449, in run
    if self.calc_h() == 0:
  File "C:\ProgramData\Anaconda3\lib\site-packages\andes\routines\tds.py", line 600, in calc_h
    if (system.dae.t + self.h) > system.switch_times[self._switch_idx]:
IndexError: index 6 is out of bounds for axis 0 with size 6

Process finished with exit code 1
```

**Reply by cuihantao** [reply]
I just pushed a change. Mind testing agin? I probably need more time to test if more thoroughly as well.

**Reply by seedpower** [reply]
The bug seems to be fixed.

The diagrams are slightly different using two methods:

```
ss = andes.run(andes.get_case("ieee14/ieee14_full.xlsx"), setup=False, default_config=True)
ss.add("Fault", dict(bus=9, tf=0.4, tc=0.7))
ss.setup()
ss.PFlow.run()

ss.TDS.config.tstep = 0.1
ss.TDS.config.shrinkt = 1

# Method 1:
for tf in np.linspace(0.1, 1.2, 12):
    ss.TDS.config.tf = tf
    ss.TDS.run()

ss.TDS.load_plotter()
ss.TDS.plotter.plot(ss.GENROU.omega, title="ss.TDS.config.tf = np.linspace(0.1, 1.2, 12)")

# Method 2:
# ss.TDS.config.tf = 1.2
# ss.TDS.run()

# ss.TDS.load_plotter()
# ss.TDS.plotter.plot(ss.GENROU.omega, title="ss.TDS.config.tf = 1.2")
```

![difference](https://user-images.githubusercontent.com/11022830/202837033-86492f23-02be-4f44-b71a-ec784bc3c240.gif)

**Reply by seedpower** [reply]
And more, a gap plot bug:
```
import numpy as np
import andes
from andes.utils import get_case

ss = andes.run(get_case(“ieee14_1.00_Line_4_0.25_1.00_1.83.xlsx”), default_config=True)

ss.TDS.config.criteria = 0

# Method 1:
for tf in np.linspace(0.1, 5.0, 50):
    print("t={t:.2f}".format(t=tf))
    ss.TDS.config.tf = tf
    ss.TDS.run()

ss.TDS.load_plotter()
ss.TDS.plotter.plot(ss.GENROU.delta)

# Method 2:
# ss.TDS.config.tf = 5.0
# ss.TDS.run()
# ss.TDS.load_plotter()
# ss.TDS.plotter.plot(ss.GENROU.delta)
```
Method 1 diagram:
![image](https://user-images.githubusercontent.com/11022830/202842820-07d0d16e-0b60-4993-a926-792fc98a97cf.png)

Method 2 diagram:
![image](https://user-images.githubusercontent.com/11022830/202842840-abd9f3fc-b55d-4dd3-8a88-f88718ef519d.png)

Existing problem:
1. discontinuity occurs in t ≈ 3.5s

case file:
[ieee14_1.00_Line_4_0.25_1.00_1.83.xlsx](https://github.com/cuihantao/andes/files/10046530/ieee14_1.00_Line_4_0.25_1.00_1.83.xlsx)

**Reply by cuihantao** [reply]
> ```
> ss = andes.run(andes.get_case("ieee14/ieee14_full.xlsx"), setup=False, default_config=True)
> ss.add("Fault", dict(bus=9, tf=0.4, tc=0.7))
> ss.setup()
> ss.PFlow.run()
> 
> ss.TDS.config.tstep = 0.1
> ss.TDS.config.shrinkt = 1
> 
> # Method 1:
> for tf in np.linspace(0.1, 1.2, 12):
>     ss.TDS.config.tf = tf
>     ss.TDS.run()
> 
> ss.TDS.load_plotter()
> ss.TDS.plotter.plot(ss.GENROU.omega, title="ss.TDS.config.tf = np.linspace(0.1, 1.2, 12)")
> 
> # Method 2:
> # ss.TDS.config.tf = 1.2
> # ss.TDS.run()
> 
> # ss.TDS.load_plotter()
> # ss.TDS.plotter.plot(ss.GENROU.omega, title="ss.TDS.config.tf = 1.2")
> ```

This one is expected since the discretization is taken at different time steps. All these points should be on the same trajectory.

**Reply by cuihantao** [reply]
> And more, a gap plot bug:
> 
> ```
> import numpy as np
> import andes
> from andes.utils import get_case
> 
> ss = andes.run(get_case(“ieee14_1.00_Line_4_0.25_1.00_1.83.xlsx”), default_config=True)
> 
> ss.TDS.config.criteria = 0
> 
> # Method 1:
> for tf in np.linspace(0.1, 5.0, 50):
>     print("t={t:.2f}".format(t=tf))
>     ss.TDS.config.tf = tf
>     ss.TDS.run()
> 
> ss.TDS.load_plotter()
> ss.TDS.plotter.plot(ss.GENROU.delta)
> 
> # Method 2:
> # ss.TDS.config.tf = 5.0
> # ss.TDS.run()
> # ss.TDS.load_plotter()
> # ss.TDS.plotter.plot(ss.GENROU.delta)
> ```
> 
> Method 1 diagram: ![image](https://user-images.githubusercontent.com/11022830/202842820-07d0d16e-0b60-4993-a926-792fc98a97cf.png)
> 
> Method 2 diagram: ![image](https://user-images.githubusercontent.com/11022830/202842840-abd9f3fc-b55d-4dd3-8a88-f88718ef519d.png)
> 
> Existing problem:
> 
> 1. discontinuity occurs in t ≈ 3.5s
> 
> case file: [ieee14_1.00_Line_4_0.25_1.00_1.83.xlsx](https://github.com/cuihantao/andes/files/10046530/ieee14_1.00_Line_4_0.25_1.00_1.83.xlsx)

Again, thank you for reporting. I pushed a fix to my `develop` branch.

The issue is due to the non-convergence at t=3.9xx. It was a bug in the step size control where the shrinked step size was not properly applied.

**Comment by cuihantao** [comment]
On a side note, if you don't need the progress bar during your reinforcement learning studies, you can turn it off by setting `ss.TDS.config.no_tqdm = 1`.

**Reply by seedpower** [reply]
Thank you, it's very useful. 

Need some help, My goal is to keep transient stability:

> In practical engineering, the generators generally are cut off as a whole.

Q1: At present, I have two ways to achieve this. One is to modify the `u` status of the generator, and the other is to set `tm0=0` of the generator . Which one is better? 

> loads usually are cut off in acertain proportion.

Q2: Can I achieve this by setting `Ppf * (1-proportion)` of PQ load?
